octodns/provider/mythicbeasts.py

```diff
@@ -0,0 +1,420 @@
+
+    @staticmethod
+    def _data_for_SSHFP(_type, data):
+        ttl = max([raw_values['ttl'] for raw_values in data['raw_values']])
+        values = []
+
+        for raw_value in \
+                [raw_values['value'] for raw_values in data['raw_values']]:
+            match = re.match(
+                '^([0-9]+)\\s+([0-9]+)\\s+(\\S+)$',
+                raw_value,
+                re.IGNORECASE)
+
+            assert match is not None, 'Unable to parse SSHFP data'
+
+            values.append({
+                'algorithm': match.group(1),
+                'fingerprint_type': match.group(2),
+                'fingerprint': match.group(3),
+            })
+
+        return {
+            'type': _type,
+            'values': values,
+            'ttl': ttl,
+        }
+
+    @staticmethod
+    def _data_for_CAA(_type, data):
+        ttl = data['raw_values'][0]['ttl']
+        raw_value = data['raw_values'][0]['value']
+
+        match = re.match(
+            '^([0-9]+)\\s+(issue|issuewild|iodef)\\s+(\\S+)$',
+            raw_value,
+            re.IGNORECASE)
+
+        assert match is not None, 'Unable to parse CAA data'
+
+        value = {
+            'flags': match.group(1),
+            'tag': match.group(2),
+            'value': match.group(3),
+        }
+
+        return MythicBeastsProvider._data_for_single(
+            'CAA',
+            {'raw_values': [{'value': value, 'ttl': ttl}]})
+
+    _data_for_NS = _data_for_multiple
+    _data_for_TXT = _data_for_multiple
+    _data_for_A = _data_for_multiple
+    _data_for_AAAA = _data_for_multiple
+
+    def populate(self, zone, target=False, lenient=False):
+        self.log.debug('populate: name=%s, target=%s, lenient=%s', zone.name,
+                       target, lenient)
+
+        resp = self.records(zone.name)
+
+        before = len(zone.records)
+        exists = False
+        data = dict()
+
+        exists = True
+        for line in resp.content.splitlines():
+            match = re.match(
+                '^(\\S+)\\s+(\\d+)\\s+(\\S+)\\s+(.*)$',
+                line,
+                re.IGNORECASE)
+
+            if match is None:
+                self.log.debug('failed to match line: %s', line)
+                continue
+
+            if match.group(1) == '@':
+                _name = ''
+            else:
+                _name = match.group(1)
+
+            _type = match.group(3)
+            _ttl = int(match.group(2))
+            _value = match.group(4).strip()
+
+            if _type == 'TXT':
+                _value = _value.replace(';', '\\;')
+
+            if hasattr(self, '_data_for_{}'.format(_type)):
+
+                if _type not in data:
+                    data[_type] = dict()
+
+                if _name not in data[_type]:
+                    data[_type][_name] = {
+                        'raw_values': [{'value': _value, 'ttl': _ttl}],
+                        'name': _name,
+                        'zone': zone.name,
+                    }
```

Comment:
data up above can use default dict to avoid the `if not in` stuff down here. It's a bit of a trade off between the complicated defaultdict bits up above and making this code down here dirt simple though.

```python
from collections import defaultdict
...
data = defaultdict(lambda: defaultdict(lambda: {
    'raw_values': [],
    'name': None,
    'zone': None,
}))
...
record = data['A']['foo']
record['name'] = _name
record['zone'] = zone.name
record['raw_values'].append({
    'value': _value,
    'ttl': _ttl,
})
```

Reply:
I'm not sure I follow this at all or the improvement?

Reply:
It's different more than a specific improvement, that'd be opinion. Personally I like that the loop doesn't have any of the `if` `else` default handling and just focuses on the logic of adding the records. The `defaultdict` stuff up top is a little bit complicated since it's dealing with two-levels, in exchange for the code in the loop just setting things directly.

Reply:
Okay no problem. I think the most recent commit should clear that up.